### PR TITLE
[#37] Transaction broadcasting additions

### PR DESCRIPTION
**Pull request by szerepak**

### CODE REVIEW

File: clove/network/base.py [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
 
-from bitcoin.core import COIN, CMutableTransaction
+from bitcoin.core import COIN, CMutableTransaction, b2lx
 from bitcoin.core.serialize import Hash, SerializationError, SerializationTruncationError
 from bitcoin.messages import (
     MSG_TX, MsgSerializable, msg_getdata, msg_inv, msg_ping, msg_pong, msg_reject, msg_tx, msg_verack, msg_version
 )
 from bitcoin.net import CInv
 
+from clove.network.exceptions import ConnectionException, TransactionRejectedException, UnexpectedResponseFromNode
```
Comment (trojkat) [inline clove/network/base.py]: Exceptions don't need _Exception_ in their names.
`ConnectionError`, `RejectedTransaction`, `UnexpectedResponse`

File: clove/utils/logging.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+def greenify(logging_message: str):
+    return f'{Fore.GREEN}{logging_message}{Style.RESET_ALL}'
+
+
+def purplify(logging_message: str):
+    return f'{Fore.MAGENTA}{logging_message}{Style.RESET_ALL}'
+
+
+def redify(logging_message: str):
```
Comment (trojkat) [inline clove/utils/logging.py]: Maybe those methods should be related to loglevel...

File: clove/network/base.py [hunk 3]
```diff
@@ -167,17 +183,34 @@ def get_current_fee_per_kb(cls) -> float:
...
-                NotImplementedError
-            return responses
+        message = msg_tx()
+        message.tx = transaction
+        self.connection.sendall(message.to_bytes())
+        logger.info(greenify(f'Transaction {transaction_hash} has just been sent.'))
+        self.connection.settimeout(20)
+
+        try:
+            responses = self.extract_all_responses(self.connection.recv(8192))
```
Comment (trojkat) [inline clove/network/base.py]: Why not 1024?

**Review by trojkat — APPROVED** [review]
```
>>> Publishing transaction
DEBUG:root:Node response messages were: 'addr, getheaders', whilst it should be 'getdata'.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
DEBUG:root:Couldn't retrieve any deserializable message.
INFO:root:Node responded correctly. Sending transaction...
INFO:root:Transaction bcbcab25569c2104e51746fdda4158aee4735d19221aa7c946079a0b89430f1a has just been sent.
>>> Transaction published!
```
I would only add a IP prefix tor those logs
```
DEBUG:root:[10.56.242.7] Node response message was: 'addr', whilst it should be 'getdata'.
```

👌

File: clove/network/base.py (hunk 3 above)
Comment (szerepak) [reply clove/network/base.py]:
```
DEBUG:root:[69.164.204.147] Couldn't retrieve any deserializable message.
DEBUG:root:[136.243.91.3] Couldn't retrieve any deserializable message.
DEBUG:root:[136.243.91.3] Couldn't retrieve any deserializable message.
DEBUG:root:[52.1.165.219] Couldn't retrieve any deserializable message.
DEBUG:root:[52.1.165.219] Couldn't retrieve any deserializable message.
DEBUG:root:[18.194.73.108] Couldn't retrieve any deserializable message.
DEBUG:root:[18.194.73.108] Couldn't retrieve any deserializable message.
DEBUG:root:[34.235.119.100] Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:[34.235.119.100] Couldn't retrieve any deserializable message.
DEBUG:root:[142.4.214.187] Couldn't retrieve any deserializable message.
DEBUG:root:[142.4.214.187] Couldn't retrieve any deserializable message.
DEBUG:root:[5.189.138.25] Couldn't retrieve any deserializable message.
DEBUG:root:[5.189.138.25] Couldn't retrieve any deserializable message.
DEBUG:root:[185.207.105.230] Couldn't retrieve any deserializable message.
DEBUG:root:[185.207.105.230] Couldn't retrieve any deserializable message.
DEBUG:root:[155.94.231.23] Couldn't retrieve any deserializable message.
DEBUG:root:[155.94.231.23] Couldn't retrieve any deserializable message.
DEBUG:root:[80.47.218.77] Node response message was: 'alert', whilst it should be 'getdata'.
DEBUG:root:[138.201.105.19] Couldn't retrieve any deserializable message.
DEBUG:root:[138.68.229.19] Couldn't retrieve any deserializable message.
DEBUG:root:[138.68.229.19] Couldn't retrieve any deserializable message.
DEBUG:root:[138.197.161.227] Couldn't retrieve any deserializable message.
DEBUG:root:[138.197.161.227] Couldn't retrieve any deserializable message.
DEBUG:root:[82.165.11.165] Couldn't retrieve any deserializable message.
DEBUG:root:[82.165.11.165] Couldn't retrieve any deserializable message.
DEBUG:root:[34.238.52.165] Couldn't retrieve any deserializable message.
DEBUG:root:[34.238.52.165] Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:[176.62.57.26] Couldn't retrieve any deserializable message.
DEBUG:root:[176.62.57.26] Couldn't retrieve any deserializable message.
DEBUG:root:[85.214.136.94] Couldn't retrieve any deserializable message.
DEBUG:root:[85.214.136.94] Couldn't retrieve any deserializable message.
DEBUG:root:[165.227.72.40] Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:[165.227.72.40] Couldn't retrieve any deserializable message.
DEBUG:root:[104.237.131.138] Couldn't retrieve any deserializable message.
DEBUG:root:[104.237.131.138] Couldn't retrieve any deserializable message.
DEBUG:root:[131.188.40.191] Couldn't retrieve any deserializable message.
DEBUG:root:[131.188.40.191] Couldn't retrieve any deserializable message.
DEBUG:root:[13.58.71.247] Couldn't retrieve any deserializable message.
DEBUG:root:[13.58.71.247] Couldn't retrieve any deserializable message.
DEBUG:root:[159.89.183.71] Couldn't retrieve any deserializable message.
DEBUG:root:[159.89.183.71] Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:[34.236.223.81] Couldn't retrieve any deserializable message.
DEBUG:root:[34.236.223.81] Couldn't retrieve any deserializable message.
DEBUG:root:[149.202.184.140] Couldn't retrieve any deserializable message.
DEBUG:root:[149.202.184.140] Node response message was: 'addr', whilst it should be 'getdata'.
DEBUG:root:[46.101.173.25] Couldn't retrieve any deserializable message.
DEBUG:root:[94.130.216.238] Couldn't retrieve any deserializable message.
DEBUG:root:[94.130.216.238] Couldn't retrieve any deserializable message.
INFO:root:[13.126.31.38] Node responded correctly. Sending transaction...
INFO:root:[13.126.31.38] Transaction cd1224df1c29c8d1677a88e0925bdec3ccde29aa32a091279e0282644e578082 has just been sent.
DEBUG:root:[13.126.31.38] Connection timeout. Node is not responding. Connection terminates.
```
It's not enough, we need more :taco: :taco:  :taco:

**Review by pafff — APPROVED** [review]

**Pull request merged**